### DISCUSSION TITLE
Github Copilot Support?

### DISCUSSION THREAD

**Discussion by TreeHD** [coder/code-server#4363] [Q&A] [answered]
Hello

I'm using code-server

as here [pull#3544](https://github.com/cdr/code-server/pull/3544#issuecomment-929712023) shows that Copilot isn't support at this moment

Is there any possible to make it usable?

**Comment by AyaanZaveri** [comment]
You can try SSHing into VSCode with [this](https://github.com/cdr/code-server/blob/main/docs/guide.md#ssh-into-code-server-on-vs-code) guide and then try to use Copilot.

Let me know if that works!

**Comment by soulmachine** [comment]
I'm having an issue using code-server with Copilot, in the `output` tab there is an error log:

```
[INFO] [auth] [2021-10-25T05:47:58.412Z] Invalid copilot token: missing token
```

**Reply by AyaanZaveri** [reply]
You need to put your OpenAI Codex API Key to use GitHub Copilot.

**Reply by AyaanZaveri** [reply]
You will need to sign up for the beta of OpenAI.

**Reply by soulmachine** [reply]
I've signed up before, and I'm able to use Copilot on my vscode of macOS

**Reply by jsjoeio** [reply]
Did this work?

**Reply by soulmachine** [reply]
No, Copilot prompted me to login to Github and I did it, but it still complained `missing token`. While I'm able to use Copilot on macOS just fine.

**Reply by Heath123** [reply]
> You will need to sign up for the beta of OpenAI.

You shouldn't need that for Copilot

**Comment by lg-nineoneninethreenine** [comment]
Guys, I have a slight doubt. Yesterday, I found a mail from GitHub, that they opted me into the Copilot Technical Preview. How can I get this on code-server? Should I get the VSIX file from the Visual Studio Marketplace, upload it onto my instance and install it via the CLI?

That kinda seems right on Gitpod, since I needed to do the same to install the Aura theme. So, does it apply the same to code-server?

**Reply by soulmachine** [reply]
I download the vsix first and use CLI to install it, see <https://github.com/soulmachine/docker-images/blob/master/code-server/Dockerfile.base#L92>

**Reply by lg-nineoneninethreenine** [reply]
Same pinch dude but can I log in to Copilot?

**Comment by Heath123** [comment]
I might try to see why this isn't working by editing/inspecting the code of Copilot and see if I can get it going

**Reply by Heath123** [reply]
I might as well share what I find as I go in case anyone has any info or wants to help:
This is the request that fails:
![image](https://user-images.githubusercontent.com/13787163/143441495-b7cf9ab1-5b39-45e2-b0c2-d5bbac4e0180.png)
I added some debug logging:
```
[INFO] [auth] [2021-11-25T12:19:28.980Z] Token: gto_[REDACTED]
[INFO] [auth] [2021-11-25T12:19:28.984Z] Response: {"message":"Not Found","documentation_url":"https://docs.github.com/rest"}
[INFO] [auth] [2021-11-25T12:19:28.984Z] Invalid copilot token: missing token
[ERROR] [default] [2021-11-25T12:19:28.987Z] GitHub Copilot could not connect to server. Extension activation failed: "User not authorized"
```
Maybe that token is somehow invalid? I should compare it to what VS Code sends and try replacing it with that

**Reply by Heath123** [reply]
Setting it to the one VS Code uses makes authentication work, but the extension itself still doesn't even though it says that it's enabled

Edit: Never mind, this happens to me in normal VS Code too, maybe something is down?

**Reply by lg-nineoneninethreenine** [reply]
Thanks a lot! I guess you aren't given the Copilot beta access, did the Copilot team give you the access?

**Reply by jfarre20** [reply]
I have beta access, and it works on normal vscode, but doesn't work on code-server
`Sign in failed: TypeError: Cannot read properties of null (reading 'blur')`

**Reply by Heath123** [reply]
> Thanks a lot! I guess you aren't given the Copilot beta access, did the Copilot team give you the access?

I do have access

**Reply by Heath123** [reply]
> I have beta access, and it works on normal vscode, but doesn't work on code-server `Sign in failed: TypeError: Cannot read properties of null (reading 'blur')`

Hmm, I just got the missing token error

**Reply by lg-nineoneninethreenine** [reply]
I too have the beta access(not even tested yet lol) but I know that installing Copilot as of now is kinda pointless. GitHub must've foolproofeed it **for now** since Copilot is in beta.

**Reply by AyaanZaveri** [reply]
> I too have the beta access(not even tested yet lol) but I know that installing Copilot as of now is kinda pointless. GitHub must've foolproofeed it **for now** since Copilot is in beta.

Yeah, also on [github.dev](https://github.dev) and [vscode.dev](https://vscode.dev) (websites created github) it says the extension is unable to be installed for the web:

<img width="781" alt="Screen Shot 2021-11-26 at 5 36 47 PM" src="https://user-images.githubusercontent.com/63752541/143659355-f2c16660-1e97-46e2-ab60-523ac5647ae4.png">

**Reply by Heath123** [reply]
The issue (or part of it) seems to be related to the fact that inline code suggestions are a proposed API, but I can't do `--enable-proposed-api` because it's saying `nullExtensionDescription` in the error message?

**Reply by Heath123** [reply]
> I too have the beta access(not even tested yet lol) but I know that installing Copilot as of now is kinda pointless. GitHub must've foolproofeed it **for now** since Copilot is in beta.

Pretty sure it's possible to work around, after all JS is relatively easy to modify compared to compiled languages, and I might get it working eventually (so far I've bypassed the login problem but it's still broken). Using print debugging I think I've isolated the problem to a block of around 5ish lines

**Reply by lg-nineoneninethreenine** [reply]
If this works, try adding a toggle and try a PR here, I'd love to see this on the repo. Also, don't forget to try creating a simple documentation for Copilot.

**Reply by jsjoeio** [reply]
> but I can't do --enable-proposed-api because it's saying nullExtensionDescription in the error message?

@code-asher do we support the `--enable-proposed-api` flag? I thought we did?

**Reply by Heath123** [reply]
It is supported I think but something is wrong with the extension ID and I don't really understand what

I've had Copilot work on normal VS Code even after showing this error so it might not be a problem

**Reply by code-asher** [reply]
We do have it but as a result of a bug if you do *not* include the flag the proposed API will be enabled for _all_ extensions by default.

**Reply by ZauberNerd** [reply]
@Heath123 try:
```sh
 sed -i 's/"GitHub.remotehub",/"GitHub.remotehub",\n    "GitHub.copilot",/g' /usr/lib/code-server/vendor/modules/code-oss-dev/product.json
```
And then use the token you extracted from regular VSCode by debugging the extension.

**Reply by ankarhem** [reply]
@ZauberNerd How can I use the token from vscode in code-server?

**Reply by ZauberNerd** [reply]
@ankarhem sorry, I can't give you detailed instructions. You'll have to debug that yourself. You could try the same as @Heath123 by editing the extension or you could probably use a MITM proxy or maybe try to extract it from some storage location or whatever.

**Reply by Heath123** [reply]
Yep, if you deminify the code you can find where it gets the token, I can help with that if you want. Sorry, I haven't really managed to try it properly yet

**Comment by Sunkist18** [comment, marked as answer]
Hi guys, I'm curious if you're still experiencing the issue because I solved it.
If you still have the issue, please refer to the following article.
Since I'm Korean, it's posted in Korean, so I'll attach the translated page as well.

### [Original Page](https://dandalf.tistory.com/112)
### [Translated Page (Eng)](https://papago.naver.net/website?locale=ko&source=auto&target=en&url=https%3A%2F%2Fdandalf.tistory.com%2F112)

**Reply by jsjoeio** [reply]
Thank you for posting it plus the translation!

**Reply by jfarre20** [reply]
I followed the steps, but it does not work. I enter the auth url and nothing happens.

**Reply by soulmachine** [reply]
It worked for me, thanks for sharing!

**Reply by Sunkist18** [reply]
> I followed the steps, but it does not work. I enter the auth url and nothing happens.

That's probably because you didn't agree to the Copilot terms and conditions.
Pop-up sometimes doesn't come up. If you refresh it, it might come up. In my case, that works.

**Reply by SamambaMan** [reply]
The extension itself won't show up at the search.
Have any of you guys been through this?

**Reply by yokowasis** [reply]
> The extension itself won't show up at the search. Have any of you guys been through this?

Because the copilot isn't available on third party extension market.

**Comment by ghost** [comment]
```
https://vscode-auth.github.com/authorize/?callbackUri=vscode://vscode.github-authentication/did-authenticate&scope=repo&state={token}&responseType=code
```

**Comment by Savid** [comment]
@Prajwal-Koirala url worked for me;

1. download latest copilot vsix
2. in code server `> Install from VSIX` and install copilot
3. when code server asks you to log in; follow that to your browser and pull out the `state` uuid token from the url
4. use @Prajwal-Koirala url and put your state token in there `https://vscode-auth.github.com/authorize/?callbackUri=vscode://vscode.github-authentication/did-authenticate&scope=repo&state={token}&responseType=code` and paste that into your browser
5. follow through to authenticate and you will eventually get to the page where it will display your URI eg. `vscode:....`. copy that
6. in code server you'll see something like "Signing into github..." status message on the bottom leftish which you can click and you paste in the URI you just got

**Reply by ghost** [reply]
Thank you for this, ur a good person.

the other guy refers u to his own website and if that site goes down so does all the info.

**Reply by lg-nineoneninethreenine** [reply]
How about documenting [this](https://github.com/coder/code-server/discussions/4363#discussioncomment-2070110)? It could make this easier.

**Reply by jsjoeio** [reply]
We'd probably need to make sure we're not breaking any legal agreements. Something tells me that Copilot can only be used in Microsoft products. However, if that's not true, we could add it to the docs.

**Reply by Savid** [reply]
they do have jetbrains/neovim [plugins](https://github.com/github/copilot-docs) at the moment. All of their current plugins are "licensed" with;
> This technical preview is a Beta Preview under the [GitHub Terms of Service](https://docs.github.com/en/github/site-policy/github-terms-of-service#j-beta-previews).

I would assume once live this would completely change as well.

**Reply by lg-nineoneninethreenine** [reply]
So, are you meaning we make a plugin for Copilot? I don't think that'd change anything, since we need to install Copilot, Copilot doesn't need to do anything. But that might seem like violating Microsoft's ToS, hence we are thinking.

**Reply by Savid** [reply]
Ahh sorry no I don't mean we make a plugin. I'm just linking what they've already made themselves and the ToS they've got. I'm not familar with Microsoft/Github's ToS and how this project and plugins relate to it.

**Reply by Heath123** [reply]
> We'd probably need to make sure we're not breaking any legal agreements. Something tells me that Copilot can only be used in Microsoft products. However, if that's not true, we could add it to the docs.

Well, it's probably against the Marketplace ToS at least

**Comment by yokowasis** [comment]
I am sorry for reopening the discussion, but the solution no longer works.

before, we can extract the state, and put it into the @Prajwal-Koirala url. Now, it doesn't work. The auth url also changed into something like "https://github.com/login?client_id=ad768f28053cf9566b63&return_to=%2Flogin%2Foauth%2Fauthorize%3Fclient_id%3Dad768f28053cf9566b63%26response_type%3Dcode%26scope%3Dread%253Auser%26state%3D%257B%2522callback%2522%253A%2522https%253A%252F%252Fsomeurl.com%252Fcallback%253Fvscode-requestId%253D80f6c042-7bd2-4aa0-8702-55351a3b14c0%255Cu0026vscode-scheme%253Dcode-oss%255Cu0026vscode-authority%253Dvscode.github-authentication%255Cu0026vscode-path%253D%25252Fdid-authenticate%2522%252C%2522state%2522%253A%252283862aae-c0b7-4e92-9631-fcc8b80d8af0%2522%257D"

I tried to decode the url to get the state, but the code-server refuse to auth and throwing bad request. 

When doing auth normaly, while it works on some extension like gitlens, it doesn't on github copilot. It says token not found [404] and it asked me to joint he waitlist.

**Reply by lg-nineoneninethreenine** [reply]
Anything in the logs?

**Reply by yokowasis** [reply]
if you mean the output
```
[INFO] [auth] [2022-03-13T01:52:07.665Z] Invalid copilot token: missing token: 404 
[ERROR] [default] [2022-03-13T01:52:07.668Z] GitHub Copilot could not connect to server. Extension activation failed: "User not authorized"
```

I am authorized, I can use it fine on desktop vs code, and could use it before using the web using @Prajwal-Koirala url. I also tried using personal token method (by using localhost) instead of url, it's the same problem.

**Reply by yokowasis** [reply]
If I tried to extract the state from the url and put it @Prajwal-Koirala URL, the output shows :
```
[ERROR] [default] [2022-03-13T01:58:42.075Z] GitHub Copilot could not connect to server. Extension activation failed: "Bad Request"
```

**Reply by lg-nineoneninethreenine** [reply]
I really don't know :(
I will try repros on a Raspberry Pi after exams tho...

**Comment by ghost** [comment]
[GitHub.copilot-1.7.5250.zip](https://github.com/coder/code-server/files/8238851/GitHub.copilot-1.7.5250.zip)

### Update 03/12/2022
### It still works.

**Reply by yokowasis** [reply]
It's not about the extension. I think, It's about the auth url. The auth url is changed. If you already auth before, it will works fine. 

Can you test it by using incognito / private window and auth the github for copilot using the latest code-server ?

https://user-images.githubusercontent.com/1374064/158049237-51e24c40-0adb-43e9-b12a-2323174b0a96.mp4

Before we still have time, to extract the state and put it into your url. Now it doesn't do that anymore.

**Reply by lg-nineoneninethreenine** [reply]
Maybe it's due to your GitHub sign-in. Can you try signing out and then try again or try on a browser that you don't use. Maybe Edge?

**Reply by lg-nineoneninethreenine** [reply]
Point is, it shouldn't have your GitHub login. Try that.

**Reply by yokowasis** [reply]
> Point is, it shouldn't have your GitHub login. Try that.

It doesn't have my github login. It's like that because I already logged in to github on that same browser. If I am not logged in, github will ask me to login, and the rest is the same. 

I have tried different browser, it's the same. I even tried cross browser. code-server in chrome, github on firefox. same result.

**Reply by lg-nineoneninethreenine** [reply]
Log out of Github on that for the time being and retry the process.

**Reply by yokowasis** [reply]
> Log out of Github on that for the time being and retry the process.

I already did that. 

> If I am not logged in, github will ask me to login, and the rest is the same.

**Reply by lg-nineoneninethreenine** [reply]
I can't debug this right now, but I will do it by Wednesday/Thursday IST. Hopefully, my browser will catch those URLs. Wait can you check your browser history for the state URI?

**Reply by yokowasis** [reply]
> I can't debug this right now, but I will do it by Wednesday/Thursday IST. Hopefully, my browser will catch those URLs. Wait can you check your browser history for the state URI?

This is the request url :
```
https://github.com/login?client_id=ad768f28053cf9566b63&return_to=%2Flogin%2Foauth%2Fauthorize%3Fclient_id%3Dad768f28053cf9566b63%26response_type%3Dcode%26scope%3Dread%253Auser%26state%3D%257B%2522callback%2522%253A%2522http%253A%252F%252F127.0.0.1%253A9090%252Fcallback%253Fvscode-requestId%253D610814c1-ea40-4de4-8909-edfdc5abb7c0%255Cu0026vscode-scheme%253Dcode-oss%255Cu0026vscode-authority%253Dvscode.github-authentication%255Cu0026vscode-path%253D%25252Fdid-authenticate%2522%252C%2522state%2522%253A%2522af75d661-fd20-46a9-843b-b081aaa26d84%2522%257D
```

this is the response url : 
```
http://127.0.0.1:9090/callback?vscode-requestId=610814c1-ea40-4de4-8909-edfdc5abb7c0&vscode-scheme=code-oss&vscode-authority=vscode.github-authentication&vscode-path=%2Fdid-authenticate&code=f556c2d5ea6a5dc55d40&state=af75d661-fd20-46a9-843b-b081aaa26d84
```

**Reply by lg-nineoneninethreenine** [reply]
Does this URL work? While having your instance open ofc
```
https://vscode-auth.github.com/authorize/?callbackUri=vscode://vscode.github-authentication/did-authenticate&scope=repo&state=af75d661-fd20-46a9-843b-b081aaa26d84&responseType=code
```

**Reply by lg-nineoneninethreenine** [reply]
Wait, maybe all you have to do is add the authorize part on your URL! Try this, if it works, then solved!
```
http://127.0.0.1:9090/authorize/callback?vscode-requestId=610814c1-ea40-4de4-8909-edfdc5abb7c0&vscode-scheme=code-oss&vscode-authority
```

**Reply by lg-nineoneninethreenine** [reply]
Sorry for the late reply. Dozed off on a Sunday.

**Reply by yokowasis** [reply]
> Does this URL work? While having your instance open ofc
> 
> ```
> https://vscode-auth.github.com/authorize/?callbackUri=vscode://vscode.github-authentication/did-authenticate&scope=repo&state=af75d661-fd20-46a9-843b-b081aaa26d84&responseType=code
> ```

Tried this before, this is what i refer to
> I tried to decode the url to get the state, but the code-server refuse to auth and throwing bad request.


> Wait, maybe all you have to do is add the authorize part on your URL! Try this, if it works, then solved!

How do you get the requestID ?,  by the time I have the requestID, auth is already successful

**Reply by lg-nineoneninethreenine** [reply]
Hmm... Maybe if you slow your webpage loading time? I don't know, need to find more insight on this.

**Reply by yokowasis** [reply]
The newest code-server 4.2.0 using a new method of auth and works fine now. The auth breaks on 4.1.0.. Version 4.0.2 and 4.2.0 works fine.

**Reply by lg-nineoneninethreenine** [reply]
Okay, great! Guess I'll just update my Pi and enjoy!

**Reply by xKliment** [reply]
Welp i tried it but i get this error from GitHub Copilot: ERROR] [default] [2023-04-25T16:45:19.666Z] GitHub Copilot could not connect to server. Extension activation failed: "Cancelled"

**Comment by benz0li** [comment]
My advice: Keep your hands off GitHub Copilot.

https://drewdevault.com/2022/06/23/Copilot-GPL-washing.html

**Reply by Heath123** [reply]
> Should a human programmer take the same approach, studying free software and applying those lessons, but not the code itself, to original projects, I would agree that their applied knowledge is not creating derivative works. However, that is not how machine learning works. **Machine learning is essentially a glorified pattern recognition and reproduction engine, and does not represent a genuine generalization of the learning process.** It is perhaps capable of a limited amount of originality, but is also capable of degrading to the simple case of copy and paste.

I disagree with the statement that machine learning isn't really learning, I feel like people clinging onto that idea just want to think that humans are special and computers will never surpass them. Really, aren't humans just glorified pattern recognition and reproduction engines? I do think it would be better if they only used permissively licensed code just to be safe, but personally I'm not going to stop using it based on the license concerns

**Comment by PawanOsman** [comment]
I used all the methods above, but I can't log in to GitHub from the copilot

I installed copilot from VSIX

I'm using the latest version of code-server
code-server: v4.8.3
Code: 1.72.1

Does anyone know how to log in to GitHub to use copilot?

**Comment by sxwebster** [comment]
Code-Server v4.9.1
Copilot version 1.65.7705 visx from marketplace

Uploaded visx to code server files pane via drag and drop. Right click install visx extension. Click on popular from copilot to log in / authorise github, follow prompts for 2fa, all signed in successfully without any issues. No copying or pasting of tokens or URLs or anything.

**Reply by yzdomino** [reply]
I can confirm this method works.
Copilot does not appear in the marketplace but does install and the authentication process works as well.

**Reply by jonahpearl** [reply]
This should be the accepted answer! Easy, and the old way doesn't seem to work anymore.

**Comment by PawanOsman** [comment]
To anyone who still can't login copilot with github accounts,

You need to remove any extension which related with github,
Its better to delete the extensions folder and restart code server then install what extension you need again,

It will fix the issue

I had this issue and i could fix it with deleting all extensions

Looks like an old extension handles the github login thats why github keep saying update your vscode

after removing all my extensions its working fine

**Comment by josephrocca** [comment]
## August 2023 update - no more tricks needed

Step-by-step process here: https://github.com/coder/code-server/discussions/5063#discussioncomment-6654481

(tl;dr: download the extension vsix file from here: https://marketplace.visualstudio.com/items?itemName=GitHub.copilot then drag and drop it into code-server filetree, then right click on it and install, then you'll get a notification to sign into github, follow the prompts and you're done)

**Reply by smerschjohann** [reply]
yes, this works for me. Did anyone succeeded with Copilot Chat as well?

**Reply by PawanOsman** [reply]
I just installed Copilot Chat and it works fine, make sure to download the latest VSIX file of Copilot Chat then extract it, modify the required vscode version to something lower than your code-server vscode version in `package.json` file in the extension directory.

for example right now latest code-server vscode version is 1.82 so i set it to ^1.80

```json
...
  "engines": {
    "vscode": "^1.80.0-20231009",
    "npm": ">=9.0.0",
    "node": ">=18.0.0"
  },
...
```

then pack the files again into a zip archive and rename the file change the extension to VSIX and done, you can upload it to your code-server and install it normally

**Reply by smerschjohann** [reply]
oh wow, thanks a lot... Never thought I could just lower the requirements of that extension :+1:

**Reply by taomanwai** [reply]
What code-server version you are using?

I am using v4.16.1, but Copilot Chat v0.11.2023111302 does not work after the installation:
Error from GitHub Copilot Chat OUTPUT tab: [extension] You are not signed in to GitHub. Please sign in to use Copilot.

My Copilot is already signed in and there is no way for me to sign in the Copilot Chat.

**Reply by smerschjohann** [reply]
I'm still on version v0.9.2023101002, that worked just fine

**Reply by taomanwai** [reply]
Could you share your v0.9.2023101002 vsix?

Both https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-chat and Google Search result do not show Copilot Chat v0.9.2023101002 vsix.

Thanks in advance.

**Reply by smerschjohann** [reply]
it works for me with latest code-server + copilot extensions as well

**Reply by nerasse** [reply]
v0.9.2023101002 was a buggy on my side and the fact that I don't have tha latest version makes me realy sad.

I did some research and I finnaly build my own code-server image with an official installation on ubuntu.
Here is my repo if you want : https://github.com/nerasse/my-code-server
gl guys

**Reply by wassname2** [reply]
`code-server --install-extension GitHub.copilot`

**Comment by Venkythota1** [comment]
Hi everyone,
I tried the above approach, when I try to sign in into the GitHub copilot. I get this pop up of You have not yet finished authorizing this extension to use GitHub. Would you like to try a different way? (personal access token) Source: GitHub Authentication when i click on yes. it asks to create a personal acess token when I get give the personal token, I am getting this below error in the output window.

auth: Error: No auth flow succeeded.
at a.GitHubServer login (/code-server-4.93.1-linux-amd64/lib/vscode/extensions/github-authentication/dist/extension js 1:228535)
at process TicksAndRejections (node internal/process/task_queues:95:5)
at a.GitHubAuthenticationProvider.createSession (/code-server-4.93.1-linux-amd64/lib/vscode/extensions/github- authentication/dist/extension.js:1:224358)
at c. $createSession (/code-server-4.93.1-linux-amd64/lib/vscode/out/vs/workbench/api/node/out-vscode-reh-
web/vs/workberich/api/node/file:/home/runner/work/code-server/code- server/lib/vscode/src/vs/workbench/api/common/extHostAuthentication.ts:89:11)

I am using the latest github copilot and vs-codeserver (v4.93.1) vs code (1.93.1)

is there any workaround for this issue?
